tracer/src/Datadog.Trace/Span.ISpan.cs

```diff
@@ -0,0 +1,77 @@
+// <copyright file="Span.ISpan.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+
+namespace Datadog.Trace
+{
+    /// <summary>
+    /// A Span represents a logical unit of work in the system. It may be
+    /// related to other spans by parent/children relationships. The span
+    /// tracks the duration of an operation as well as associated metadata in
+    /// the form of a resource name, a service name, and user defined tags.
+    /// </summary>
+    public partial class Span : ISpan
+    {
+        /// <inheritdoc />
+        string ISpan.OperationName
```

Comment:
I think this is only required because `Span` is a public type but all of the members we want to expose in `ISpan` are `internal` when accessed through the object. I do think it would be simpler to make `Span` internal and then we don't need all of these explicit implementations. What do you all think?

Reply:
The introduction of the explicit interface implementations was to avoid an API that looked like this:

```csharp
public class Span
{
    public ISpanContext Context => Context;
    internal SpanContext InternalContext { get; }
}
```

It makes for a more confusing API for us - when should we use `Context` vs `InternalContext`? Does that change with mixed versions rewriting etc? Obviously they're the same _currently_ but not guaranteed to be.

Making the interface explicit, and given we work with internal APIs, means we can rename `InternalContext` -> `Context`, and means we always use the correct one ("Internal" one for auto instrumentation, "interface" one for manual instrumentation).

The reason for implementing _all_ of the interface explicitly, is that converting a property from "normally" implemented to explicitly implemented is a breaking change for duck typing. By making everything explicit now, we are free to change the implementations of the internal `Span` methods without breaking backwards compatibility.

As for the final point, whether `Span` can be `internal`. Good question 🤔 I don't think it needs to be `public` any more, but I'm not 100% on that. It's an easy change to make later though, so I don't think it should block us merging this?

Reply:
That was really helpful, thanks for the detailed explanation. I agree that we could push the `Span` public/internal to a later PR if we decide we want to change this

Reply:
Thanks, Lock!